src/main/java/rx/observables/SyncOnSubscribe.java

```diff
@@ -337,14 +331,39 @@ private SubscriptionProducer(final Subscriber<? super T> subscriber, SyncOnSubsc
 
         @Override
         public boolean isUnsubscribed() {
-            return isUnsubscribed != 0;
+            return get() < 0L;
         }
         
         @Override
         public void unsubscribe() {
-            IS_UNSUBSCRIBED.compareAndSet(this, 0, 1);
-            if (get() == 0L)
-                parent.onUnsubscribe(state);
+            while(true) {
```

Comment:
There is a much simpler way of fixing this: undo all these latest changes and request 1!

```java
if (BackpressureUtils.addRequested(this, 1) == 0) {
    parent.onUnsubscribe(state);
}
```